torch/csrc/jit/tensorexpr/eval.h

```diff
@@ -368,11 +387,31 @@ class SimpleIREvaluator : public CodeGen, public IRVisitor {
     if (lhs_v.dtype() != rhs_v.dtype()) {
       throw malformed_input("bad dtype in binary op", v);
     }
+
     IRNodeType expr_type = v->expr_type();
     if (expr_type == IRNodeType::kAnd || expr_type == IRNodeType::kOr ||
-        expr_type == IRNodeType::kXor || expr_type == IRNodeType::kLshift ||
-        expr_type == IRNodeType::kRshift) {
-      value_ = bitwise_binary_op(lhs_v, rhs_v, expr_type);
+        expr_type == IRNodeType::kXor) {
+      switch (lhs_v.dtype().scalar_type()) {
+#define TYPE_CASE(Type, Name)                                  \
+  case ScalarType::Name:                                       \
+    value_ = bitwise_binary_op<Type>(lhs_v, rhs_v, expr_type); \
+    break;
+        AT_FORALL_INT_TYPES_AND(Bool, TYPE_CASE);
+#undef TYPE_CASE
+        default:
+          throw unsupported_dtype();
+      }
+      return;
+    }
+
+    if (expr_type == IRNodeType::kLshift || expr_type == IRNodeType::kRshift) {
+      switch (lhs_v.dtype().scalar_type()) {
+        case ScalarType::Int:
```

Comment:
But we now need to support different dtypes here as well rather than always using the `int` version, right? I think we need to use `AT_FORALL_INT_TYPES` here as well.

Btw, we don't need to add `_AND` to the macro if the only reason for that is to support the bool case in the bitwise ops - we can add an explicit `case` for `ScalarType::bool` instead.

Reply:
in `ir.h`, the constructor of `Lshift` IR seems to only support operands of type `ScalarType::Int`. 

```
class Lshift : public BinaryOpNode<Lshift> {
 public:
  Lshift(const Expr* lhs, const Expr* rhs)
      : BinaryOpNode(lhs, rhs, IRNodeType::kLshift) {
    if (lhs->dtype().scalar_type() != ScalarType::Int) {
      throw unsupported_dtype();
    }
    if (lhs->dtype() != rhs->dtype()) {
      throw malformed_input("bad dtype in Lshift");
    }
  }
};
```

I was trying to make the implementation of `Lshift` and `Rshift` to be consistent w/ its IR definition. But supporting other integer types (which also makes sense for these IRs) in the implementation will not cause any trouble. So I am fine either way. It's your call.

As for `ScalarType::bool`, I will add an explicit case switch for it to make it more readable.

Reply:
> I was trying to make the implementation of Lshift and Rshift to be consistent w/ its IR definition. But supporting other integer types (which also makes sense for these IRs) in the implementation will not cause any trouble. So I am fine either way. It's your call.

Ultimately the ground truth here is the semantics of aten ops that we're representing in tensor expressions. In case of shifts the following script could serve as a reference for us:
```
import torch

dtypes = [
  torch.uint8,
  torch.int8,
  torch.int16,
  torch.int32,
  torch.float16,
  torch.float32,
  torch.float64,
  torch.bool,
]

for dev in ['cpu', 'cuda']:
    for dt in dtypes:
        x = torch.ones(4, dtype=dt, device=dev)
        try:
            y = x << 2
            print(y)
        except Exception as e:
            print(e)
```

If we run it, we'll see that the dtypes should be preserved through these operations:
```
tensor([4, 4, 4, 4], dtype=torch.uint8)
tensor([4, 4, 4, 4], dtype=torch.int8)
tensor([4, 4, 4, 4], dtype=torch.int16)
tensor([4, 4, 4, 4], dtype=torch.int32)
"lshift_cpu" not implemented for 'Half'
tensor([4., 4., 4., 4.])
tensor([4., 4., 4., 4.], dtype=torch.float64)
"lshift_cpu" not implemented for 'Bool'
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.uint8)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int8)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int16)
tensor([4, 4, 4, 4], device='cuda:0', dtype=torch.int32)
tensor([4., 4., 4., 4.], device='cuda:0', dtype=torch.float16)
tensor([4., 4., 4., 4.], device='cuda:0')
tensor([4., 4., 4., 4.], device='cuda:0', dtype=torch.float64)
"lshift_cuda" not implemented for 'Bool'
```

But as you noted, this is a slightly bigger change and seems like a separate issue to me. Please feel free to land this PR and fix behavior of shifts later in a different one.

Reply:
Btw, I'm adding tests that would expose this discrepancy between the eager and TE-compiled behavior of shift operations in #48292.